browser/src/statechart.js

```diff
@@ -0,0 +1,615 @@
+};
+
+const autofitState = {
+  on: {
+    click: { actions: 'autofit' },
+    TOGGLERGB: {
+      target: 'paint',
+      in: '#deepcellLabel.rgb.oneChannel'
+    },
+  }
+};
+
+const watershedState = {
+  entry: 'storeClick',
+  initial: 'idle',
+  states: {
+    idle: {
+      on: {
+        click: {
+          cond: 'validSeed',
+          target: 'clicked',
+          actions: 'storeClick'
+        }
+      }
+    },
+    clicked: {
+      on: {
+        click: {
+          cond: 'validSecondSeed',
+          actions: 'watershed'
+        }
+      }
+    }
+  },
+  on: {
+    TOGGLERGB: {
+      target: 'paint',
+      in: '#deepcellLabel.rgb.oneChannel'
+    },
+  }
+};
+
+/**
+ * Handles which current tool for mouse behavior.
+ */
+const toolbarState = {
+  initial: 'paint',
+  states: {
+    flood: floodState,
+    trim: trimState,
+    erodeDilate: erodeDilateState,
+    autofit: autofitState,
+    paint: paintState,
+    threshold: thresholdState,
+    watershed: watershedState,
+    hist: { type: 'history' }
+  },
+  on: {
+    LOADING: 'loading',
+    'keydown.b': '.paint',
+    'keydown.g': '.flood',
+    'keydown.k': '.trim',
+    'keydown.q': '.erodeDilate',
+    // tools not available in RGB mode
+    'keydown.t': {
+      target: '.threshold',
+      in: '#deepcellLabel.rgb.oneChannel'
+    },
+    'keydown.m': {
+      target: '.autofit',
+      in: '#deepcellLabel.rgb.oneChannel'
+    },
+    'keydown.w': {
+      target: '.watershed',
+      in: '#deepcellLabel.rgb.oneChannel'
+    },
+  }
+};
+
+/**
+ * Handles tool behavior with a loading state.
+ */
+const mouseState = {
+  initial: 'toolbar',
+  states: {
+    toolbar: toolbarState,
+    loading: {
+      on: { LOADED: 'toolbar.hist' }
+    }
+  },
+  on: {
+    EDIT: { actions: forwardTo('backend') },
+    UNDO: { actions: forwardTo('backend') },
+    REDO: { actions: forwardTo('backend') },
+  }
+};
+
+/**
+ * Handles adjustments to the images displayed on the canvas.
+ */
+const adjusterState = {
+  on: {
+    SCROLLCONTRAST: { actions: 'changeContrast' },
+    SCROLLBRIGHTNESS: { actions: 'changeBrightness' },
+    'keydown.h': { actions: 'toggleHighlight' },
+    'keydown.0': { actions: 'resetBrightnessContrast' },
+    'keydown.i': { actions: 'toggleInvert' }
+  }
+};
+
+/**
+ * Handles moving around the canvas.
+ */
+const canvasState = {
+  initial: 'idle',
+  states: {
+    idle: {
+      on: { 'keydown.space': 'pan' },
+    },
+    pan: {
+      on: {
+        'keyup.space': 'idle',
+        mousemove: { actions: ['pan', 'updateMousePos'] }
+      }
+    }
+  },
+  on: {
+    'keydown.-': { actions: send({ type: 'ZOOM', change: -1 }) },
+    'keydown.=': { actions: send({ type: 'ZOOM', change: 1 }) },
+    ZOOM: { actions: 'zoom' },
+    mousemove: { actions: 'updateMousePos' }
+  }
+};
+
+const brushState = {
+  on: {
+    'keydown.up': { actions: 'incrementBrushSize' },
+    'keydown.down': { actions: 'decrementBrushSize' }
+  }
+};
+
+const displayState = {
```

Comment:
<img width="621" alt="Screen Shot 2021-02-16 at 2 18 16 PM" src="https://user-images.githubusercontent.com/40068392/108130979-65641e00-7065-11eb-8bbf-b630ee203d6f.png">

The display state keeps track of whether we are viewing the raw image, the labels, or the overlay.